### PR TITLE
add c++ solution to guard mark (gold)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/usaco-494.mdx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+Let's go through a dynamic programming approach.
+
+#### States
+
+We have to keep track of the cows that we're using ($mask$).
+
+To check if we've surpassed Mark's height already and to return the result, we should also keep track of the current height ($h$) and the maximum safety factor achievable ($sf$).
+
+$\texttt{dp}[mask] =\{h, sf\}$
```
Comment (envyaims) [inline solutions/usaco-494.mdx:31]: assigning $h$ and $sf$ and this line isn't necessary

File: solutions/usaco-494.mdx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+#### Base Cases
+
+With no cows, the safety factor will be infinite and the height will be zero.
+
+#### Transitions
+
+$\texttt{dp}[mask] = \max_{j \in mask}(\texttt{dp}[mask],\hspace{0.1cm}\min(\texttt{dp}[prev] - weight_j,\hspace{0.1cm}strength_j))$
+
+Such that $j =$ the $j$th cow and $weight_j,\hspace{0.1cm}strength_j$ represents the weight and strength of cow $j$ respectively.
```
Comment (envyaims) [inline solutions/usaco-494.mdx:41]: what does this mean? why are there $j$ in both sides of the equation lol

File: solutions/usaco-494.mdx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+$\texttt{dp}[mask] =\{h, sf\}$
+
+#### Base Cases
+
+With no cows, the safety factor will be infinite and the height will be zero.
+
+#### Transitions
+
+$\texttt{dp}[mask] = \max_{j \in mask}(\texttt{dp}[mask],\hspace{0.1cm}\min(\texttt{dp}[prev] - weight_j,\hspace{0.1cm}strength_j))$
```
Comment (envyaims) [inline solutions/usaco-494.mdx:39]: whats $prev$?

**Review by envyaims — CHANGES_REQUESTED** [review]

File: solutions/usaco-494.mdx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+author: Ryan Chou
+---
+ 
+[Official Analysis](http://www.usaco.org/current/data/sol_guard_gold.html)
+
+## Explanation
+
+Note that since it might not always be optimal to keep the strongest/heaviest cow at the bottom of the stack, we have to go through every single possible subset, which is feasible with $N \leq 20$.
+
+example of a counter case:
```
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/usaco-494.mdx:14]:
```suggestion
Here's an example of a counter case:
```

File: solutions/usaco-494.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-494.mdx:39]: also don't use lovecraftian math formulas please

File: solutions/usaco-494.mdx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			ans = max(dp[i].second, ans);
+		}
+	}
+
+	// no tall enough/stable arrangement exists
+	if (ans < 0) {
+		cout << "Mark is too tall" << endl;
+	} else {
+		cout << ans << endl;
+	}
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:113]: ternary expression

File: solutions/usaco-494.mdx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+				* maximum safety factor if we put cow j
+				* at the top of our stack (account for j's strength as well)
+				*/
+				dp[i].second = max(dp[i].second,
+                                   min(dp[prev].second - cows[j].weight, cows[j].strength));
+			}
+		}
+	}
+
+	int ans = -1;
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:101]: rename to `max_safety` maybe?

File: solutions/usaco-494.mdx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+				// update our height
+				dp[i].first += cows[j].height;
+
+				int prev = i ^ (1 << j);
+				/*
+				* maximum safety factor if we put cow j
+				* at the top of our stack (account for j's strength as well)
+				*/
+				dp[i].second = max(dp[i].second,
+                                   min(dp[prev].second - cows[j].weight, cows[j].strength));
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:96]:
```suggestion
					dp[i].second = max(
						dp[i].second,
						min(dp[prev].second - cows[j].weight, cows[j].strength)
					);
```

File: solutions/usaco-494.mdx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	int n;
+	int h;
+	cin >> n >> h;
+
+	vector<Cow> cows(n);
+	for (int i = 0; i < n; i++) {
+		cin >> cows[i].height >> cows[i].weight >> cows[i].strength;
+	}
+
+	// dp[i] = {height, safety factor}
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:75]: it's been a while since i did this problem
is there only 1 height?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-494.mdx (hunk 8 above)
Comment (ryanchou-dev) [reply solutions/usaco-494.mdx:75]: ill rename for clarity

File: solutions/usaco-494.mdx (hunk 5 above)
Comment (dongliu0426) [reply solutions/usaco-494.mdx:115]: I don't think ternary expressions can return different types

File: solutions/usaco-494.mdx (hunk 3 above)
Comment (ryanchou-dev) [reply solutions/usaco-494.mdx]: err, i explain it in the next-next line

Comment (ryanchou-dev) [reply solutions/usaco-494.mdx]: ok one sec

File: solutions/usaco-494.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/usaco-494.mdx]: oh yeah, nvm

File: solutions/usaco-494.mdx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+Here's an example of a counter case:
+```
+4 6
+1 1000 4
+1 1 5
+3 2 3
+3 3 3
+```
+
+Let's go through a dynamic programming approach.
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:23]: something like
"hm bounds are 20"
"99% of the time that's bitmask dp so let's try that & see what happens"

File: solutions/usaco-494.mdx [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+#### States
+
+We have to keep track of the cows that we're using ($mask$).
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:27]:
```suggestion
Let's use a bitmask ($mask$) to keep track of the cows we're using.
```

File: solutions/usaco-494.mdx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+To check if we've surpassed Mark's height already and to return the result, we should also keep track of the current height and the best safety factor we can get.
+
+$\texttt{dp}[mask] =\{$current height, maximum safety factor achievable$\}$
+
+#### Base Cases
```
Comment (SansPapyrus683) [inline solutions/usaco-494.mdx:33]:
```suggestion
#### Base Case
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
almost done

File: solutions/usaco-494.mdx (hunk 9 above)
Comment (ryanchou-dev) [reply solutions/usaco-494.mdx:23]: ok sure

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/usaco-494.mdx [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+3 3 3
+```
+
+Almost every time the bounds are ~$20$, it suggests bitmask DP.
+
+Let's go through the information we need to implement it!
+
+#### States
+
+Let's use a bitmask ($mask$) to keep track of the cows we're using.
```
Comment (maggie-j-liu) [inline solutions/usaco-494.mdx:29]: use `\texttt{}` for var names longer than 1 letter

File: solutions/usaco-494.mdx [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+				int prev = i ^ (1 << j);
+				/*
+				* maximum safety factor if we put cow j
+				* at the top of our stack (account for j's strength as well)
+				*/
+					dp[i].second = max(
+						dp[i].second,
+						min(dp[prev].second - cows[j].weight, cows[j].strength)
+					);
```
Comment (maggie-j-liu) [inline solutions/usaco-494.mdx:100]:
```suggestion
				dp[i].second = max(
					dp[i].second,
					min(dp[prev].second - cows[j].weight, cows[j].strength)
				);
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-494.mdx [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+Here's an example of a counter case:
+```
+4 6
+1 1000 4
+1 1 5
+3 2 3
+3 3 3
+```
+
```
Comment (envyaims) [inline solutions/usaco-494.mdx:22]: a one sentence+ explanation would be nice even though it’s obvious

**Review by envyaims — APPROVED** [review]
yayy!

File: solutions/usaco-494.mdx [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+Almost every time the bounds are ~$20$, it suggests bitmask DP.
```
Comment (envyaims) [inline solutions/usaco-494.mdx:23]:
```suggestion
Almost every time the bounds are ~$20$, it suggests looping over subsets.
```

File: solutions/usaco-494.mdx [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+Almost every time the bounds are ~$20$, it suggests looping over subsets.
+
+Let's go through the information we need to implement it!
+
+#### States
+
+Let's use a bitmask ($\texttt{mask}$) to keep track of the cows we're using.
+
+To check if we've surpassed Mark's height already and to return the result, we should also keep track of the current height and the best safety factor we can get.
```
Comment (jjliewie) [inline solutions/usaco-494.mdx:32]:
```suggestion
To check if we've surpassed Mark's height already and if so, return the result, we should also keep track of the current height and the best safety factor we can get.
```

**Review by jjliewie — APPROVED** [review]

**Pull request merged**